### ISSUE TITLE
stack read underflow in tparam_internal

### ISSUE THREAD

**Issue by hannob** [vim/vim#2638]
Address Sanitizer detects a stack read underflow in vim for me.

This happens only if
TERM=xterm-256color
is set as an environment variable.

I tried to analyze the code to find the bug, but find it confusing. I hope someone more familiar can find it.
This was tested on a full asan-enabled system. I was unable to manually reproduce this in an asan build on another system, yet I thought I'd still report it. The stack trace involves ncurses, so it *might* be that the bug also involves ncurses or even is a bug in ncurses.

Full address sanitizer stack trace:
```
==583==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ffcafb06b30 at pc 0x7fb982af6168 bp 0x7ffcafb068a0 sp 0x7ffcafb06890
READ of size 8 at 0x7ffcafb06b30 thread T0
    #0 0x7fb982af6167 in tparam_internal /var/tmp/portage/sys-libs/ncurses-6.0-r1/work/ncurses-6.0/ncurses/tinfo/lib_tparm.c:513
    #1 0x7fb982af6167 in tparm /var/tmp/portage/sys-libs/ncurses-6.0-r1/work/ncurses-6.0/ncurses/tinfo/lib_tparm.c:834
    #2 0x7def6a in term_color /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/term.c:2847
    #3 0x71bb77 in screen_start_highlight /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8094
    #4 0x72246a in screen_start_highlight /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8359
    #5 0x72246a in screen_char /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8313
    #6 0x72630b in screen_fill /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8648
    #7 0x727b7d in win_draw_end /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:2431
    #8 0x743891 in win_update /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:2273
    #9 0x7495fa in update_screen /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:756
    #10 0x882036 in main_loop /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/main.c:1245
    #11 0x883071 in vim_main2 /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/main.c:905
    #12 0x430d73 in main /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/main.c:429
    #13 0x7fb9819b15ac in __libc_start_main (/lib64/libc.so.6+0x205ac)
    #14 0x432979 in _start (/usr/bin/vim+0x432979)

Address 0x7ffcafb06b30 is located in stack of thread T0 at offset 0 in frame
    #0 0x7deebf in term_color /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/term.c:2821

  This frame has 1 object(s):
    [32, 52) 'buf'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-underflow /var/tmp/portage/sys-libs/ncurses-6.0-r1/work/ncurses-6.0/ncurses/tinfo/lib_tparm.c:513 tparam_internal
Shadow bytes around the buggy address:
  0x100015f58d10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100015f58d20: 00 00 00 00 f1 f1 f1 f1 04 f4 f4 f4 f2 f2 f2 f2
  0x100015f58d30: 00 00 00 f4 f2 f2 f2 f2 00 00 00 00 00 00 00 00
  0x100015f58d40: 00 f4 f4 f4 f2 f2 f2 f2 00 00 00 00 00 00 00 00
  0x100015f58d50: 00 f4 f4 f4 f3 f3 f3 f3 00 00 00 00 00 00 00 00
=>0x100015f58d60: 00 00 00 00 00 00[f1]f1 f1 f1 00 00 04 f4 f3 f3
  0x100015f58d70: f3 f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100015f58d80: 00 00 00 00 00 00 f1 f1 f1 f1 00 00 06 f4 f3 f3
  0x100015f58d90: f3 f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100015f58da0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100015f58db0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Contiguous container OOB:fc
  ASan internal:           fe
==583==ABORTING
```

**Comment by chrisbra** [comment]
If I am not mistaken, this looks like a bug in ncurses. Can you reproduce it with ncurses-6.1 or another 6.0 version?

**Comment by brammool** [comment]
Hanno Böck wrote:

> Address Sanitizer detects a stack read underflow in vim for me.
> 
> This happens only if
> TERM=xterm-256color
> is set as an environment variable.

Can you check the output of ":set termcap" in this situation?

> I tried to analyze the code to find the bug, but find it confusing. I
> hope someone more familiar can find it.
> This was tested on a full asan-enabled system. I was unable to
> manually reproduce this in an asan build on another system, yet I
> thought I'd still report it. The stack trace involves ncurses, so it
> *might* be that the bug also involves ncurses or even is a bug in
> ncurses.
> 
> Full address sanitizer stack trace:
> ```
> ==583==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ffcafb06b30 at pc 0x7fb982af6168 bp 0x7ffcafb068a0 sp 0x7ffcafb06890
> READ of size 8 at 0x7ffcafb06b30 thread T0
>     #0 0x7fb982af6167 in tparam_internal /var/tmp/portage/sys-libs/ncurses-6.0-r1/work/ncurses-6.0/ncurses/tinfo/lib_tparm.c:513
>     #1 0x7fb982af6167 in tparm /var/tmp/portage/sys-libs/ncurses-6.0-r1/work/ncurses-6.0/ncurses/tinfo/lib_tparm.c:834
>     #2 0x7def6a in term_color /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/term.c:2847

What is the code in this line?  I suppose it's a call to tgoto(), but
which one?

>     #3 0x71bb77 in screen_start_highlight /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8094
>     #4 0x72246a in screen_start_highlight /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8359
>     #5 0x72246a in screen_char /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8313
>     #6 0x72630b in screen_fill /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:8648
>     #7 0x727b7d in win_draw_end /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:2431
>     #8 0x743891 in win_update /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:2273
>     #9 0x7495fa in update_screen /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/screen.c:756
>     #10 0x882036 in main_loop /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/main.c:1245
>     #11 0x883071 in vim_main2 /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/main.c:905
>     #12 0x430d73 in main /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/main.c:429
>     #13 0x7fb9819b15ac in __libc_start_main (/lib64/libc.so.6+0x205ac)
>     #14 0x432979 in _start (/usr/bin/vim+0x432979)
> 
> Address 0x7ffcafb06b30 is located in stack of thread T0 at offset 0 in frame
>     #0 0x7deebf in term_color /var/tmp/portage/app-editors/vim-8.0.1428/work/vim-8.0.1428/src/term.c:2821
> 
>   This frame has 1 object(s):
>     [32, 52) 'buf'
> HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext

Sounds like the printf-like handling of the termcap entry is done wrong.

I do notice that the "i" variable here is always two, instead of setting
to one when using 8-bit termcap strings.  It's been like that for a very
long time, but it is wrong.


-- 
ARTHUR:        I command you as King of the Britons to stand aside!
BLACK KNIGHT:  I move for no man.
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by hannob** [comment]
I tested with ncurses 6.1 and this no longer happens, so maybe it indeed was an ncurses bug.
I saw there've been plenty of memory safety fixes in 6.1. From my side this is resolved, but it seems @brammool identified issues in the code nevertheless that should be fixed?

**Comment by chrisbra** [comment]
closing then.
